### DISCUSSION TITLE
Normalization of CLIP Vectors Before Storing in Milvus for Cosine Similarity

### DISCUSSION THREAD

**Discussion by Bruc3Stark** [milvus-io/milvus#32479] [Q&A and General discussion] [answered]
After vectors are transformed by the CLIP model, do they need to be normalized before being stored in Milvus? If they are stored in the database without normalization, will it affect the results when calculating cosine similarity?

**Comment by yhmo** [comment, marked as answer]
There are three metric types for floating-point-vectors to calculate vector similarity: IP, L2 and COSINE: https://milvus.io/docs/metric.md#Similarity-Metrics

When you call create_index(), you can specify a metric type.
```
index_params = {
    'metric_type': "COSINE",     # metric type
    'index_type': "IVF_FLAT",    # index type
    'params': {"nlist": 128},         # index parameters
}
collection.create_index(field_name=VECTOR_FIELD_NAME, index_params=index_params)
```

If your vectors are not normalized, and you choose IP metric type, the result will be incorrect, because IP metric requires vectors must be normalized.

If your vectors are not normalized, you can choose the COSINE metric type, Milvus will automatically calculate the similarity with normalized vectors, the result is correct.

If your vectors are not normalized, and you choose L2 metric type, the result is also correct because L2 is Euclidean distance, it doesn't require vectors to be normalized.


Note: the results of COSINE and L2 for un-normalized vectors are different, although both of them can handle un-normalized vectors because the mathematical meaning is different.

**Reply by ChendiWang** [reply]
I have a question on if the vectors need to be normalized for the IP metric calculation part. 

To my understanding, you can get more information on comparing the vector's magnitude by not normalizing the vector when using the IP metric. - Related reference from Milvus documentation: https://milvus.io/docs/metric.md?tab=floating#:~:text=IP%20is%20more%20useful%20if%20you%20need%20to%20compare%20non%2Dnormalized%20data%20or%20when%20you%20care%20about%20magnitude%20and%20angle.

If that's the case, would  the combination of (non-normalized vectors, IP metric) be recommended?

**Reply by xiaofan-luan** [reply]
it's kind of based on your model.
Each model I guess have a recommended way of  using what kind of metrics.
Usually L2 metrics is usually a generalized way for multi modal

**Reply by ChendiWang** [reply]
Thanks for your reply.

More specifically, I'm wondering what kind of incorrect results will incur based on the quoted part of the original reply: 
"If your vectors are not normalized, and you choose IP metric type, the result will be incorrect, because IP metric requires vectors must be normalized."

Seems like IP metric calculation allows for non-normalized vector input based on this definition of IP from Milvus documentation?: https://milvus.io/docs/metric.md?tab=floating#:~:text=the%20distance%20metric.-,Inner%20product%20(IP),-The%20IP%20distance

**Reply by xiaofan-luan** [reply]
no, ip distance don't need to be normalized. and if you do normalized on ip instance then it becomes cosine distance

**Reply by ChendiWang** [reply]
Thanks a lot for your confirmation!
